RavenDB-22246: Introduce an option in TestDriver to throw if there is no license or license could not be activated

ArieSLV opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-22246

### Additional description

Implementation of `LicenseOptions` within `ServerOptions`, including the six previously existing configurations:
- `License`
- `LicensePath`
- `EulaAccepted`
- `DisableAutoUpdate`
- `DisableAutoUpdateFromApi`
- `DisableLicenseSupportCheck`

Plus the introduction of a new configuration: `EnforceLicense`.

If `EnforceLicense` is set to true, the test or embedded server will not start if the license is not provided in `LicenseOptions` or if the provided license fails validation.

### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [x] New feature

### How risky is the change?

- [x] Low 
- [ ] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [ ] Non breaking change
- [x] Ensured:
The `ServerOptions.AcceptEula` option has been marked as obsolete and moved inside `ServerOptions.LicenseConfiguration` as `EulaAccepted`. Backward compatibility is achieved as follows:

```csharp
public bool AcceptEula
{
    get => LicenseConfiguration.EulaAccepted;
    set => LicenseConfiguration.EulaAccepted = value;
}
```
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [x] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [ ] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

ArieSLV committed:
RavenDB-22246: Introduce an option in TestDriver to throw if there is no license or license could not be activated

ppekrol commented on src/Raven.Embedded/ServerOptions.cs:
```diff
@@ -92,7 +94,18 @@ public class SecurityOptions
+            public string License { get; set; }
+            public string LicensePath { get; set; }
+            public bool EulaAccepted { get; set; }
+            public bool DisableAutoUpdate { get; set; }
```
maybe this in 5.4 can be true as well?

ArieSLV replied:
I think it's better to leave the default value for `DisableAutoUpdate` property unchanged for both version 5.4 and 6.0. I would prefer not to break the license update logic during the server version upgrade process.

ppekrol commented on src/Raven.Embedded/ServerOptions.cs:
```diff
@@ -92,7 +94,18 @@ public class SecurityOptions
+            public string LicensePath { get; set; }
+            public bool EulaAccepted { get; set; }
+            public bool DisableAutoUpdate { get; set; }
+            public bool DisableAutoUpdateFromApi { get; set; }
```
maybe this in 5.4 can be true as well?

ArieSLV replied:
I think it's better to leave the default value for `DisableAutoUpdateFromApi` property unchanged for both version 5.4 and 6.0. I would prefer not to break the license update logic during the server version upgrade process.

ppekrol commented on src/Raven.Embedded/ServerOptions.cs:
```diff
@@ -92,7 +94,18 @@ public class SecurityOptions
+            public bool EulaAccepted { get; set; }
+            public bool DisableAutoUpdate { get; set; }
+            public bool DisableAutoUpdateFromApi { get; set; }
+            public bool DisableLicenseSupportCheck { get; set; }
```
maybe this in 5.4 can be true as well?

ArieSLV replied:
Done.

ppekrol commented on src/Raven.Embedded/ServerOptions.cs:
```diff
@@ -92,7 +94,18 @@ public class SecurityOptions
+            public bool DisableAutoUpdate { get; set; }
+            public bool DisableAutoUpdateFromApi { get; set; }
+            public bool DisableLicenseSupportCheck { get; set; }
+            public bool EnforceLicense { get; set; }
```
maybe ThrowOnLicenseActivationFailure ?

ArieSLV replied:
I was thinking that it might not be clear enough that an exception will also be thrown if there's no license at all. No license means nothing to activate, so it might seem like there wouldn't be an activation failure and no exception.

How about we go with `ThrowOnInvalidOrMissingLicense` instead? This way, it's clear that an exception will be thrown if the license is invalid or missing.

ppekrol replied:
works for me

ArieSLV replied:
Done.

ppekrol commented on src/Raven.Server/RavenServer.cs:
```diff
@@ -153,7 +155,10 @@ public RavenServer(RavenConfiguration configuration)
+            BeforeSchemaUpgrade = VerifyLicense;
+
+            if (Configuration.Licensing.EnforceLicense)
+                AfterDatabaseCreation = VerifyLicense;
```
why this is needed? can you describe me the flow?

ArieSLV replied:
We have two scenarios for running an Embedded server with `EnforceLicense == true`:

1. Starting a fresh server without a System database.
2. Starting a previously used server with an existing System database.

In the first case, we perform the license verification at the final stage of `LoadExistingDatabase` before starting `UpgradeSchemaIfRequired`.

In the second case, the verification is done immediately after creating the System database.

I could rename the `BeforeSchemaUpgrade` delegate to something like `VerifyLicenseAction` and use it in both cases...

ppekrol replied:
Leave it, it is fine

ArieSLV committed:
RavenDB-22246:
- `EnforceLicense` rename to `ThrowOnInvalidOrMissingLicense`
- `DisableLicenseSupportCheck ` default is `true`

grisha-kotler commented on src/Raven.Server/Commercial/LicenseHelper.cs:
```diff
@@ -86,6 +88,20 @@ public License TryGetLicenseFromString(bool throwOnFailure)
+            if (licenseStatus.Expiration.HasValue)
+                expirationDate = licenseStatus.Expiration.Value;
+
+            return expirationDate >= RavenVersionAttribute.Instance.ReleaseDate;
```
Why not use the `Expired` field here?

ArieSLV replied:
That's our initial approach (I haven't altered it): we want to allow the server to start even if the license has expired, and we don't want to permit updating RavenDB to a version newer than the license expiration date.

grisha-kotler replied:
Ok
